File: apps/control-plane-backend/control_plane_backend/self_test/corpus.py
from __future__ import annotations

from pydantic import BaseModel

# The golden corpus is crafted so retrieval has ONE provable answer: a unique
# marker phrase lives in exactly one library. We assert on *which* document is
# retrieved, never on LLM prose — so the check is deterministic even against a
# real vector store. See ADMIN-SELF-TEST-HARNESS-RFC §4.

# A nonsense, collision-free marker so a keyword/semantic hit is unambiguous.
MARKER_PHRASE = "Marchtober"


class TestDocument(BaseModel):
    filename: str
    text: str


class TestLibrary(BaseModel):
    # Stable name lets seeding be idempotent (reconcile by name, never duplicate).
    name: str
    description: str
    document: TestDocument


# Library ALPHA holds the marker fact; BETA holds only unrelated content.
LIBRARY_ALPHA = TestLibrary(
    name="fred-selftest-alpha",
    description="Self-test golden corpus (alpha) — safe to delete.",
    document=TestDocument(
        filename="fred-selftest-alpha.md",
        text=(
            "# Fred self-test fixture document ALPHA\n\n"
            f"The Fredchurro festival takes place in {MARKER_PHRASE}.\n\n"
            "This sentence is the only place the marker fact appears."
        ),
    ),
)

LIBRARY_BETA = TestLibrary(
    name="fred-selftest-beta",
    description="Self-test golden corpus (beta) — safe to delete.",
    document=TestDocument(
        filename="fred-selftest-beta.md",
        text=(
            "# Fred self-test fixture document BETA\n\n"
            "This document is about unrelated topics: weather, gardening, and tea.\n\n"
            "It deliberately contains no festival information at all."
        ),
    ),
)

# The probe question used to validate scoping. Scoped to ALPHA it must surface
# the marker; scoped to BETA it must not.
PROBE_QUESTION = "When does the Fredchurro festival take place?"

File: apps/frontend/src/rework/features/pipeline/scenarios/corpus.ts
// Copyright Thales 2026
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// The golden corpus: the document text AND its provable answer live together,
// so retrieval has one unambiguous outcome. The marker appears in exactly one
// library, which makes "scoped to A → found" / "scoped to B → absent" a
// deterministic assertion even though a real RAG stack answered.

export const MARKER = "Marchtober";
// A second marker that lives only in BETA. The isolation turn (festival question
// scoped to B) must RETRIEVE B content — proven by echoing this marker — so that
// "MARKER absent" actually means "isolated", not "B retrieval returned nothing".
export const BETA_MARKER = "Gardenbruary";
export const PROBE = "When does the Fredchurro festival take place?";
export const SELF_TEST_AGENT_ID = "fred.github.self_test";

// The prompt journeys assert *delivery* of a prompt the agent echoes back. The
// agent prints `system_prompt: …` / `context_prompt: …`; the harness injects a
// run-scoped marker into each and asserts it comes back. The base marker words
// below are suffixed with the run token in the scenario so they never collide.
export const SYSTEM_PROMPT_MARKER_BASE = "FredSelftestSysprompt";
export const CONTEXT_PROMPT_MARKER_BASE = "FredSelftestCtxprompt";
export const PROMPT_PROBE = "Self-test: confirm the delivered prompts.";

export interface CorpusDoc {
  library: string;
  fileName: string;
  text: string;
}

export const ALPHA: CorpusDoc = {
  library: "fred-selftest-alpha",
  fileName: "fred-selftest-alpha.md",
  text:
    "# Fred self-test fixture ALPHA\n\n" +
    `The Fredchurro festival takes place in ${MARKER}.\n\n` +
    "This sentence is the only place the marker fact appears.\n",
};

export const BETA: CorpusDoc = {
  library: "fred-selftest-beta",
  fileName: "fred-selftest-beta.md",
  text:
    "# Fred self-test fixture BETA\n\n" +
    `This document is about unrelated topics: weather, gardening, and tea, filed under the codeword ${BETA_MARKER}.\n\n` +
    "It deliberately contains no festival information at all.\n",
};

/** Synthesize the upload file from the inline corpus text. */
export function fileOf(doc: CorpusDoc): File {
  return new File([doc.text], doc.fileName, { type: "text/markdown" });
}
